File: examples/python/x402_signer_client.py
"""
M2M Sentinel — Headless x402 Python Signer Client.

Handles HTTP 402 Payment Required challenges by signing EIP-712 / EIP-3009
transfer authorizations for Base USDC without browser wallet dependencies.
"""

import json
import base64
import time
import os
import urllib.request
import urllib.error

BASE_USDC_CONTRACT = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
BASE_CHAIN_ID = 8453


EXPECTED_PAYOUT_RECIPIENT = "0x6d6c398390cfb88f1cd42715b84906a0bd6652aa"
DEFAULT_MAX_PRICE_USD = 0.05  # 5 cents maximum per autonomous request


class X402SignerClient:
    def __init__(self, private_key=None, base_url="https://api.m2msentinel.com", timeout=30, max_price_usd=DEFAULT_MAX_PRICE_USD, expected_recipient=EXPECTED_PAYOUT_RECIPIENT):
        self.private_key = private_key or os.getenv("OPERATOR_TEST_WALLET_KEY")
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_price_usd = float(max_price_usd)
        self.max_amount_units = int(self.max_price_usd * 1_000_000)
        self.expected_recipient = expected_recipient

    def _parse_challenge(self, header_value, body):
        if header_value:
            try:
                return json.loads(header_value)
            except Exception:
                try:
                    return json.loads(base64.b64decode(header_value).decode("utf-8"))
                except Exception:
                    pass
        if isinstance(body, dict):
            if "accepts" in body and len(body["accepts"]) > 0:
                return body["accepts"][0]
            if "paymentRequired" in body:
                return body["paymentRequired"]
        return None

    def sign_authorization(self, challenge):
        if not self.private_key:
            raise ValueError("Signer private key is required to sign x402 payment authorization")

        pay_to = challenge.get("payTo") or challenge.get("recipient") or self.expected_recipient
        amount_units = str(challenge.get("maxAmountRequired") or challenge.get("amountUnits") or "5000")
        token_contract = challenge.get("assetContract") or BASE_USDC_CONTRACT
        chain_id = int(challenge.get("chainId") or BASE_CHAIN_ID)

        # Local Security Invariant Checks
        if chain_id != BASE_CHAIN_ID:
            raise ValueError(f"[x402 Security Policy] Refusing to sign on unverified network chainId: {chain_id}. Expected Base Mainnet (8453).")
        if token_contract.lower() != BASE_USDC_CONTRACT.lower():
            raise ValueError(f"[x402 Security Policy] Refusing to sign for unapproved asset: {token_contract}. Expected Base USDC ({BASE_USDC_CONTRACT}).")
        if pay_to.lower() != self.expected_recipient.lower():
            raise ValueError(f"[x402 Security Policy] Refusing to sign for unexpected recipient: {pay_to}. Expected {self.expected_recipient}.")
        if int(amount_units) > self.max_amount_units:
            raise ValueError(f"[x402 Security Policy] Requested amount ({amount_units} units) exceeds local client authorized price ceiling ({self.max_amount_units} units).")

        now = int(time.time())
        valid_after = now - 60
        valid_before = now + 3600
        nonce = "0x" + os.urandom(32).hex()

        try:
            from eth_account import Account
            from eth_account.messages import encode_typed_data

            account = Account.from_key(self.private_key)
            from_address = account.address

            structured_data = {
                "types": {
                    "EIP712Domain": [
                        {"name": "name", "type": "string"},
                        {"name": "version", "type": "string"},
                        {"name": "chainId", "type": "uint256"},
                        {"name": "verifyingContract", "type": "address"},
                    ],
                    "TransferWithAuthorization": [
                        {"name": "from", "type": "address"},
                        {"name": "to", "type": "address"},
                        {"name": "value", "type": "uint256"},
                        {"name": "validAfter", "type": "uint256"},
                        {"name": "validBefore", "type": "uint256"},
                        {"name": "nonce", "type": "bytes32"},
                    ],
                },
                "primaryType": "TransferWithAuthorization",
                "domain": {
                    "name": "USD Coin",
                    "version": "2",
                    "chainId": chain_id,
                    "verifyingContract": token_contract,
                },
                "message": {
                    "from": from_address,
                    "to": pay_to,
                    "value": int(amount_units),
                    "validAfter": valid_after,
                    "validBefore": valid_before,
                    "nonce": bytes.fromhex(nonce[2:]),
                },
            }

            signable = encode_typed_data(full_message=structured_data)
            signed = account.sign_message(signable)
            signature = "0x" + signed.signature.hex()

            return {
                "x402Version": 2,
                "scheme": "eip3009",
                "network": "base",
                "chainId": chain_id,
                "asset": "USDC",
                "assetContract": token_contract,
                "authorization": {
                    "from": from_address,
                    "to": pay_to,
                    "value": amount_units,
                    "validAfter": valid_after,
                    "validBefore": valid_before,
                    "nonce": nonce,
                    "v": signed.v,
                    "r": "0x" + signed.r.to_bytes(32, "big").hex(),
                    "s": "0x" + signed.s.to_bytes(32, "big").hex(),
                    "signature": signature,
                },
            }
        except ImportError:
            # Fallback envelope for mock / dry-run tests when eth-account is not in environment
            return {
                "x402Version": 2,
                "scheme": "eip3009",
                "network": "base",
                "chainId": chain_id,
                "asset": "USDC",
                "assetContract": token_contract,
                "authorization": {
                    "from": "0x0000000000000000000000000000000000000001",
                    "to": pay_to,
                    "value": amount_units,
                    "validAfter": valid_after,
                    "validBefore": valid_before,
                    "nonce": nonce,
                    "signature": "0x" + "00" * 65,
                },
            }

    def request(self, endpoint_path, method="GET", body=None, headers=None):
        url = f"{self.base_url}/{endpoint_path.lstrip('/')}"
        req_headers = {
            "Accept": "application/json",
            "User-Agent": "M2MSentinel-PythonX402/1.1.0",
        }
        if headers:
            req_headers.update(headers)

        data = json.dumps(body).encode("utf-8") if body else None
        req = urllib.request.Request(url, data=data, headers=req_headers, method=method)

        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                resp_body = json.loads(resp.read().decode("utf-8"))
                return {"statusCode": resp.status, "body": resp_body, "headers": dict(resp.headers)}
        except urllib.error.HTTPError as e:
            if e.code == 402:
                resp_data = e.read().decode("utf-8")
                try:
                    resp_body = json.loads(resp_data)
                except Exception:
                    resp_body = {"raw": resp_data}

                challenge_header = e.headers.get("payment-required") or e.headers.get("x-payment-required")
                challenge = self._parse_challenge(challenge_header, resp_body)
                if not challenge:
                    raise RuntimeError("HTTP 402 received with no valid payment challenge")

                payment_payload = self.sign_authorization(challenge)
                encoded = base64.b64encode(json.dumps(payment_payload).encode("utf-8")).decode("utf-8")

                req_headers["x-payment-response"] = encoded
                paid_req = urllib.request.Request(url, data=data, headers=req_headers, method=method)
                with urllib.request.urlopen(paid_req, timeout=self.timeout) as paid_resp:
                    paid_body = json.loads(paid_resp.read().decode("utf-8"))
                    return {
                        "statusCode": paid_resp.status,
                        "body": paid_body,
                        "headers": dict(paid_resp.headers),
                        "paymentPayload": payment_payload,
                    }
            raise


if __name__ == "__main__":
    print("M2M Sentinel Headless Python x402 Signer ready.")

File: examples/x402_signer_client.js
'use strict';

/**
 * Facilitator-Independent Headless x402 Signer Client.
 *
 * Implements automated HTTP 402 challenge negotiation and EIP-712 / EIP-3009
 * transfer authorization signing directly on Base mainnet (chainId: 8453).
 * Zero browser or UI wallet dependencies — pure headless operation for autonomous agents.
 */

const https = require('https');
const http = require('http');
const crypto = require('crypto');
const { ethers } = require('ethers');

const BASE_USDC_CONTRACT = '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913';
const BASE_CHAIN_ID = 8453;

function parsePaymentHeader(value) {
  if (!value) return null;
  if (typeof value === 'object') return value;
  try { return JSON.parse(value); } catch (_) {}
  try {
    const normalized = String(value).replace(/-/g, '+').replace(/_/g, '/');
    const padded = normalized + '='.repeat((4 - normalized.length % 4) % 4);
    return JSON.parse(Buffer.from(padded, 'base64').toString('utf8'));
  } catch (_) {
    return null;
  }
}

function parsePriceToUnits(priceStr, decimals = 6) {
  if (typeof priceStr === 'number') return BigInt(Math.round(priceStr * 10 ** decimals));
  const clean = String(priceStr).replace(/[^0-9.]/g, '');
  const [whole, fraction = ''] = clean.split('.');
  const paddedFraction = (fraction + '0'.repeat(decimals)).slice(0, decimals);
  return BigInt(whole || '0') * BigInt(10 ** decimals) + BigInt(paddedFraction);
}

const EXPECTED_PAYOUT_RECIPIENT = '0x6d6c398390cfb88f1cd42715b84906a0bd6652aa';
const DEFAULT_MAX_PRICE_USD = 0.05; // 5 cents maximum per autonomous request
const EIP712_TOKEN_NAME = 'USD Coin';
const EIP712_TOKEN_VERSION = '2';

class X402SignerClient {
  constructor(options = {}) {
    this.wallet = options.wallet || (options.privateKey ? new ethers.Wallet(options.privateKey) : null);
    this.baseUrl = options.baseUrl || 'https://api.m2msentinel.com';
    this.timeoutMs = Number(options.timeoutMs || 30000);
    this.expectedRecipient = EXPECTED_PAYOUT_RECIPIENT;
    this.maxPriceUsd = options.maxPriceUsd !== undefined ? Number(options.maxPriceUsd) : DEFAULT_MAX_PRICE_USD;
    this.maxAmountUnits = parsePriceToUnits(this.maxPriceUsd, 6);
  }

  async signPaymentAuthorization(challenge = {}) {
    if (!this.wallet) {
      throw new Error('Signer wallet is required to sign x402 payment authorization');
    }

    // 1. IMMUTABLE LOCAL SECURITY CONSTANTS (Never challenge-controlled)
    const chainId = BASE_CHAIN_ID; // Base Mainnet (8453)
    const tokenContract = BASE_USDC_CONTRACT; // Base USDC (0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913)
    const payTo = EXPECTED_PAYOUT_RECIPIENT; // M2M Sentinel Payout (0x6d6c398390cfb88f1cd42715b84906a0bd6652aa)

    // x402 v2 nests the offer under `accepts`. Reading the flat shape alone made
    // every lookup below miss, so the integrity checks passed vacuously and the
    // amount silently fell back to a hardcoded default -- signing an
    // authorization for less than the server demanded, which is refused. Read
    // the offer first, then keep the flat shape as a fallback for older servers.
    const offer = (Array.isArray(challenge.accepts) && challenge.accepts[0]) || challenge;
    const offeredChainId = typeof offer.network === 'string' && offer.network.startsWith('eip155:')
      ? Number(offer.network.slice('eip155:'.length))
      : (offer.chainId || challenge.chainId);
    // v2 carries the address in `asset`; the legacy shape carries a symbol there
    // and the address in `assetContract`. Only ever compare address to address,
    // so a symbol cannot be mistaken for a mismatched contract.
    const offeredAsset = [offer.assetContract, challenge.assetContract, offer.asset, challenge.asset]
      .find((value) => typeof value === 'string' && /^0x[0-9a-fA-F]{40}$/.test(value)) || null;
    const offeredPayTo = offer.payTo || offer.recipient || challenge.payTo;
    const offeredTokenName = (offer.extra && offer.extra.name) || challenge.tokenName;
    const offeredTokenVersion = (offer.extra && offer.extra.version) || challenge.tokenVersion;

    // 2. STRICT CHALLENGE INTEGRITY CHECKS (Refuse if challenge alters network, asset, or recipient)
    if (offeredChainId && Number(offeredChainId) !== BASE_CHAIN_ID) {
      throw new Error(`[x402 Security Policy] Refusing to sign on unverified network chainId: ${offeredChainId}. Autonomous signer strictly requires Base Mainnet (8453).`);
    }
    if (offeredAsset && offeredAsset.toLowerCase() !== BASE_USDC_CONTRACT.toLowerCase()) {
      throw new Error(`[x402 Security Policy] Refusing to sign for unapproved asset: ${offeredAsset}. Autonomous signer strictly requires Base USDC (${BASE_USDC_CONTRACT}).`);
    }
    if (offeredPayTo && offeredPayTo.toLowerCase() !== EXPECTED_PAYOUT_RECIPIENT.toLowerCase()) {
      throw new Error(`[x402 Security Policy] Refusing to sign for unexpected recipient: ${offeredPayTo}. Autonomous signer strictly requires ${EXPECTED_PAYOUT_RECIPIENT}.`);
    }
    if (offeredTokenName && offeredTokenName !== EIP712_TOKEN_NAME) {
      throw new Error(`[x402 Security Policy] Refusing to sign for unexpected tokenName: ${offeredTokenName}. Expected ${EIP712_TOKEN_NAME}.`);
    }
    if (offeredTokenVersion && offeredTokenVersion !== EIP712_TOKEN_VERSION) {
      throw new Error(`[x402 Security Policy] Refusing to sign for unexpected tokenVersion: ${offeredTokenVersion}. Expected ${EIP712_TOKEN_VERSION}.`);
    }

    // 3. STRICT LOCAL PRICE CEILING CHECK
    // Never invent a price. Signing a guessed amount produces an authorization
    // the server refuses, which is indistinguishable from a rejected payment and
    // was exactly the failure this client shipped with.
    //
    // v2 states base units ("20000"); the legacy shape states a price ("$0.005").
    // Both are read, neither is assumed.
    const rawAmount = offer.amount ?? offer.maxAmountRequired ??
      challenge.maxAmountRequired ?? challenge.amountUnits ?? challenge.amount ?? challenge.price;
    let requestedAmountUnits = null;
    if (typeof rawAmount === 'number' && Number.isInteger(rawAmount) && rawAmount > 0) {
      requestedAmountUnits = String(rawAmount);
    } else if (typeof rawAmount === 'string' && /^\d+$/.test(rawAmount.trim())) {
      requestedAmountUnits = rawAmount.trim();
    } else if (typeof rawAmount === 'string' && /[$.]/.test(rawAmount)) {
      requestedAmountUnits = parsePriceToUnits(rawAmount, 6).toString();
    }
    if (!requestedAmountUnits || !/^\d+$/.test(requestedAmountUnits) || BigInt(requestedAmountUnits) <= 0n) {
      throw new Error('[x402] The challenge carried no readable amount. Refusing to guess a price.');
    }
    if (BigInt(requestedAmountUnits) > this.maxAmountUnits) {
      throw new Error(`[x402 Security Policy] Requested amount (${requestedAmountUnits} units) exceeds local client authorized price ceiling (${this.maxAmountUnits.toString()} units / $${this.maxPriceUsd}).`);
    }
    const amountUnits = requestedAmountUnits;

    // x402 v2 is identified by the presence of `accepts`. The two protocol
    // versions need different envelopes, and sending a v1 envelope to a v2
    // server is silently unpayable, so the version is decided once, here.
    const isV2 = Array.isArray(challenge.accepts) && challenge.accepts.length > 0;

    const now = Math.floor(Date.now() / 1000);
    // v2 servers bound the authorization window with `maxTimeoutSeconds`. A
    // longer window than the server advertised is not more permissive, it is
    // simply outside what the facilitator will settle.
    const validAfter = isV2 ? '0' : now - 60;
    const validBefore = isV2
      ? String(now + (Number(offer.maxTimeoutSeconds) || 120))
      : now + 3600;
    const nonce = '0x' + crypto.randomBytes(32).toString('hex');

    const domain = {
      name: EIP712_TOKEN_NAME,
      version: EIP712_TOKEN_VERSION,
      chainId,
      verifyingContract: tokenContract
    };

    const types = {
      TransferWithAuthorization: [
        { name: 'from', type: 'address' },
        { name: 'to', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'validAfter', type: 'uint256' },
        { name: 'validBefore', type: 'uint256' },
        { name: 'nonce', type: 'bytes32' }
      ]
    };

    const message = {
      from: await this.wallet.getAddress(),
      to: payTo,
      value: amountUnits.toString(),
      validAfter,
      validBefore,
      nonce
    };

    let signature;
    if (typeof this.wallet.signTypedData === 'function') {
      signature = await this.wallet.signTypedData(domain, types, message);
    } else {
      signature = await this.wallet._signTypedData(domain, types, message);
    }

    const sigParts = ethers.Signature.from(signature);

    if (isV2) {
      // The canonical v2 envelope. `accepted` is the offer echoed back
      // verbatim: the server matches it against its own requirements with a
      // deep equality check to learn *which* offer is being paid. Omitting it
      // does not read as "no payment" -- it crashes the server's matcher, which
      // answers 402 with an opaque body. Reconstructing the offer instead of
      // echoing it fails the same way, so `offer` is passed through untouched.
      //
      // Note there is deliberately no top-level `scheme`/`network` here: in v2
      // they live inside `accepted`, and the payload schema does not carry them.
      return {
        x402Version: 2,
        resource: challenge.resource,
        accepted: offer,
          // Echo the server's advertised extensions back. The canonical client
          // merges paymentRequired.extensions into the payload; omitting the
          // field still settles the payment, so this failed silently -- but the
          // CDP Bazaar indexes a resource from the bazaar extension carried by
          // the settlement, so a payment without it is invisible to discovery.
        extensions: challenge.extensions,
        payload: {
          authorization: {
            from: message.from,
            to: message.to,
            value: message.value,
            validAfter,
            validBefore,
            nonce
          },
          signature
        }
      };
    }

    // Legacy flat envelope, for servers that do not advertise `accepts`.
    return {
      x402Version: 2,
      scheme: 'eip3009',
      network: 'base',
      chainId,
      asset: 'USDC',
      assetContract: tokenContract,
      authorization: {
        from: message.from,
        to: message.to,
        value: message.value,
        validAfter,
        validBefore,
        nonce,
        v: sigParts.v,
        r: sigParts.r,
        s: sigParts.s,
        signature
      }
    };
  }

  async request(endpointPath, options = {}) {
    const url = new URL(endpointPath, this.baseUrl);
    const isHttps = url.protocol === 'https:';
    const transport = isHttps ? https : http;

    const perform = (headers = {}) => new Promise((resolve, reject) => {
      const allHeaders = {
        Accept: 'application/json',
        'User-Agent': 'M2MSentinel-X402Signer/1.1.0',
        ...options.headers,
        ...headers
      };

      const req = transport.request(url, {
        method: options.method || 'GET',
        headers: allHeaders,
        timeout: this.timeoutMs
      }, (res) => {
        let data = '';
        res.setEncoding('utf8');
        res.on('data', (chunk) => { data += chunk; });
        res.on('end', () => {
          let body = null;
          if (data) {
            try { body = JSON.parse(data); } catch (_) { body = { raw: data }; }
          }
          resolve({
            statusCode: res.statusCode,
            headers: res.headers,
            body
          });
        });
      });

      req.on('timeout', () => req.destroy(new Error('Request timeout')));
      req.on('error', reject);
      if (options.body) {
        req.write(typeof options.body === 'string' ? options.body : JSON.stringify(options.body));
      }
      req.end();
    });

    // Step 1: initial request
    const firstRes = await perform();

    // If 200 OK or not 402, return immediately
    if (firstRes.statusCode !== 402) {
      return firstRes;
    }

    // Step 2: Extract payment challenge
    const challengeHeader = firstRes.headers['payment-required'] || firstRes.headers['x-payment-required'];
    let challenge = parsePaymentHeader(challengeHeader);
    if (!challenge && firstRes.body && (firstRes.body.accepts || firstRes.body.paymentRequired)) {
      challenge = firstRes.body.accepts ? firstRes.body.accepts[0] : firstRes.body.paymentRequired;
    }

    if (!challenge) {
      throw new Error('HTTP 402 received but no valid payment challenge was present');
    }

    // Step 3: Sign payment authorization
    const paymentPayload = await this.signPaymentAuthorization(challenge);
    const encodedPaymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString('base64');

    // Step 4: Re-submit request with payment authorization
    const paidRes = await perform({
      // x402 v2 uses PAYMENT-SIGNATURE for the signed retry. The old
      // payment-response names are response headers and are ignored by the
      // gateway, which made this otherwise-correct example unpayable.
      'PAYMENT-SIGNATURE': encodedPaymentHeader
    });

    paidRes.paymentPayload = paymentPayload;
    return paidRes;
  }
}

module.exports = {
  X402SignerClient,
  BASE_USDC_CONTRACT,
  BASE_CHAIN_ID,
  parsePaymentHeader,
  parsePriceToUnits
};

if (require.main === module) {
  console.log('M2M Sentinel Headless x402 Signer Client ready.');
}
